wincode/src/schema/containers.rs

```diff
@@ -0,0 +1,596 @@
+///     address: Box<[Address]>
+/// }
+///
+/// let my_struct = MyStruct {
+///     address: vec![Address(array::from_fn(|i| i as u8)); 10].into_boxed_slice(),
+/// };
+/// let wincode_bytes = wincode::serialize(&my_struct).unwrap();
+/// let bincode_bytes = bincode::serialize(&my_struct).unwrap();
+/// assert_eq!(wincode_bytes, bincode_bytes);
+/// # }
+/// ```
+#[cfg(feature = "alloc")]
+pub struct BoxedSlice<T, Len = BincodeLen>(PhantomData<T>, PhantomData<Len>);
+
+#[cfg(feature = "alloc")]
+/// Like [`BoxedSlice`], for [`Rc`].
+pub struct RcSlice<T, Len = BincodeLen>(PhantomData<T>, PhantomData<Len>);
+
+#[cfg(feature = "alloc")]
+/// Like [`BoxedSlice`], for [`Arc`].
+pub struct ArcSlice<T, Len = BincodeLen>(PhantomData<T>, PhantomData<Len>);
+
+/// Indicates that the type is an element of a sequence, composable with [`containers`](self).
+///
+/// Prefer [`Pod`] for types representable as raw bytes.
+pub struct Elem<T>(PhantomData<T>);
+
+/// Indicates that the type is represented by raw bytes, composable with sequence [`containers`](self)
+/// or compound types (structs, tuples) for an optimized read/write implementation.
+///
+/// Use [`Elem`] with [`containers`](self) that aren't comprised of POD.
+///
+/// This can be useful outside of sequences as well, for example on newtype structs
+/// containing byte arrays / vectors with `#[repr(transparent)]`.
+///
+/// # Examples
+///
+/// A repr-transparent newtype struct with a byte array:
+/// ```
+/// # #[cfg(all(feature = "alloc", feature = "derive"))] {
+/// # use wincode::{containers::{self, Pod}};
+/// # use wincode_derive::{SchemaWrite, SchemaRead};
+/// # use serde::{Serialize, Deserialize};
+/// # use std::array;
+/// #[derive(Serialize, Deserialize)]
+/// #[repr(transparent)]
+/// struct Address([u8; 32]);
+///
+/// #[derive(Serialize, Deserialize, SchemaWrite, SchemaRead)]
+/// struct MyStruct {
+///     #[wincode(with = "Pod<_>")]
+///     address: Address
+/// }
+///
+/// let my_struct = MyStruct {
+///     address: Address(array::from_fn(|i| i as u8)),
+/// };
+/// let wincode_bytes = wincode::serialize(&my_struct).unwrap();
+/// let bincode_bytes = bincode::serialize(&my_struct).unwrap();
+/// assert_eq!(wincode_bytes, bincode_bytes);
+/// # }
+/// ```
+pub struct Pod<T>(PhantomData<T>);
+
+impl<T> SchemaWrite for Pod<T> {
+    type Src = T;
+
+    #[inline]
+    fn size_of(_src: &Self::Src) -> Result<usize> {
+        Ok(size_of::<T>())
+    }
+
+    #[inline]
+    fn write(writer: &mut Writer, src: &Self::Src) -> Result<()> {
+        // SAFETY: `T` is plain ol' data.
+        unsafe { writer.write_t(src) }
+    }
+}
+
+impl<T> SchemaRead<'_> for Pod<T> {
+    type Dst = T;
+
+    fn read(reader: &mut Reader, dst: &mut MaybeUninit<Self::Dst>) -> Result<()> {
+        // SAFETY: `T` is plain ol' data.
+        unsafe { reader.read_t(dst) }
+    }
+}
+
+#[cfg(feature = "alloc")]
+impl<T, Len> SchemaWrite for Vec<Elem<T>, Len>
```

Comment:
if Pod was a trait, we also wouldn't need a separate schema impl for the alloc
containers I think?